src/server/pkg/obj/cache_client.go

```diff
@@ -0,0 +1,149 @@
+package obj
+
+import (
+	"context"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	log "github.com/sirupsen/logrus"
+)
+
+var _ Client = &cacheClient{}
+
+type cacheClient struct {
+	slow, fast Client
+
+	mu           sync.Mutex
+	cache        *simplelru.LRU
+	populateOnce sync.Once
+}
+
+// NewCacheClient returns slow wrapped in an LRU write-through cache using fast for storing
+// cached data.
+func NewCacheClient(slow, fast Client, size int) Client {
+	if size == 0 {
+		return slow
+	}
+	client := &cacheClient{
+		slow: slow,
+		fast: fast,
+	}
+	cache, err := simplelru.NewLRU(size, client.onEvicted)
+	if err != nil {
+		// lru.NewWithEvict only errors for size < 1
+		panic(err)
+	}
+	client.cache = cache
+	return client
+}
+
+func (c *cacheClient) Reader(ctx context.Context, p string, offset, size uint64) (io.ReadCloser, error) {
+	c.doPopulateOnce(ctx)
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, exists := c.cache.Get(p); exists {
+		return c.fast.Reader(ctx, p, offset, size)
+	}
+	if err := Copy(ctx, c.slow, c.fast, p, p); err != nil {
+		return nil, err
+	}
+	c.cache.Add(p, struct{}{})
+	return c.fast.Reader(ctx, p, offset, size)
+}
+
+func (c *cacheClient) Writer(ctx context.Context, p string) (io.WriteCloser, error) {
+	wc, err := c.slow.Writer(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+	return &cacheWriter{WriteCloser: wc, path: p, client: c}, nil
+}
+
+func (c *cacheClient) Delete(ctx context.Context, p string) error {
+	if err := c.slow.Delete(ctx, p); err != nil {
+		return err
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.deleteFromCache(ctx, p)
+}
+
+func (c *cacheClient) Exists(ctx context.Context, p string) bool {
+	c.doPopulateOnce(ctx)
+	return c.fast.Exists(ctx, p) || c.slow.Exists(ctx, p)
```

Comment:
I think we should only be checking the slow client here. We depend on this to determine if we should upload a chunk or not. (We reserve a chunk with gc to prevent deletion, then use the existence check to determine if we need to upload, so a stale cache entry could be problematic here). Really, I think we should primarily be leaning on the fast client for read operations. The error type checks probably make sense as is though.

Reply:
We always write directly to the slow layer (write through). Anything that makes it into the fast layer, had to have come from the slow layer.  We call Exists a lot, and it also benefits from reduced latency. I think we should keep it as is.

Reply:
A different Pachd instance could have uploaded the chunk being checked. We cannot depend on the cache for this check. This is operating on full chunks, so the reduced latency does not get us much and could result with an additional write if the chunk exists. The real gain from the cache will be doing many small reads from a chunk (multiple small files in a chunk).

Reply:
If a different pachd uploaded the chunk then `fast` returns false and `slow` is evaluated.  So if we don't have the chunk in fast we check both.  There is no reason not to do this; It occasionally saves a `slow.Exists` and it isn't even an extra line of code.

Reply:
Whoops, my bad. The problem is with deletes not uploads. If a Pachd instance deletes a chunk and it is cached in a different Pachd instance, then the Pachd instance with it cached will think it exists in the slow storage when it does not.